File: Bike_Rental_Count.py
#!/usr/bin/env python
# coding: utf-8

# ## Importing Required libraries

# In[271]:


import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from sklearn.model_selection import train_test_split,RandomizedSearchCV
from sklearn.metrics import mean_squared_error
from math import sqrt
import statsmodels.api as sm
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor


# In[272]:


# Loading the dataset
data = pd.read_csv("day.csv")


# In[273]:


#Checking the dimensions of dataset
data.shape


# In[274]:


data.head()


# In[275]:


#Inital insight of datset
data.describe()


# In[276]:


data.dtypes


# In[277]:


## Removing unnessary variables from dataset
# instant - It is basically index number
# dteday - All the values from dteday are present in datset under differnet variables
# casual and registered - cnt is basically the sum of casual amd registerd variables
data = data.drop(['instant','dteday','casual','registered'],axis=1)


# In[278]:


#Creating a copy of original dataset
data_vis = data.copy()
data.head()


# In[279]:


##Converting the interger values into proper naming
data_vis['season'] = data_vis['season'].replace([1,2,3,4],['Springer','summer','fall','winter'])
data_vis['yr'] = data_vis['yr'].replace([0,1],[2011,2012])
data_vis['weathersit'] = data_vis['weathersit'].replace([1,2,3,4],[' Clear+Few clouds+Partly cloudy','Mist + Cloudy, Mist + Broken clouds, ',' Light Snow, Light Rain + Thunderstorm ','Heavy Rain + Ice Pallets '])
data_vis['holiday'] = data_vis['holiday'].replace([0,1],['working Day','Holiday'])
data_vis.head()


# In[280]:


print(data.dtypes)
print(data.head())


# ## Univarient Analysis

# In[281]:


## Bar Graph for Categorical data

sns.set_style("whitegrid")
sns.factorplot(data=data_vis,x='season',kind='count',size=4,aspect=2)
sns.factorplot(data=data_vis,x='yr',kind='count',size=4,aspect=2)
sns.factorplot(data=data_vis,x='mnth',kind='count',size=4,aspect=2)
sns.factorplot(data=data_vis,x='holiday',kind='count',size=4,aspect=2)
sns.factorplot(data=data_vis,x='workingday',kind='count',size=4,aspect=2)
sns.factorplot(data=data_vis,x='weathersit',kind='count',size=4,aspect=2)


# In[282]:


plt.hist(data_vis['temp'],bins=30)
plt.xlabel('temp')
plt.ylabel('Frequency')
plt.show()


# In[283]:


plt.hist(data_vis['atemp'],bins=30)
plt.xlabel('atemp')
plt.ylabel('Frequency')
plt.show()


# In[284]:


plt.hist(data_vis['hum'],bins=30)
plt.xlabel('humidity')
plt.ylabel('Frequency')
plt.show()


# In[285]:


plt.hist(data_vis['windspeed'],bins=30)
plt.xlabel('WindSpeed')
plt.ylabel('Frequency')
plt.show()


# In[286]:


#Dimensions of dataset after removing outliers
data.shape


# ## Bivariant Analysis

# In[287]:


##  Using Scatter Plot
# Index(['instant', 'season', 'yr', 'mnth', 'holiday', 'weekday', 'workingday',
#        'weathersit', 'temp', 'atemp', 'hum', 'windspeed', 'casual',
#        'registered', 'cnt'],
#       dtype='object')


# In[288]:


fig,x = plt.subplots(nrows= 2,ncols=2)
fig.set_size_inches(12,15)
sns.scatterplot(x="temp",y = "cnt",data = data_vis,palette="Set3",ax=x[0][0])
sns.scatterplot(x="atemp",y = "cnt",data = data_vis,palette="Set3",ax=x[0][1])
sns.scatterplot(x="hum",y = "cnt",data = data_vis,palette="Set3",ax=x[1][0])
sns.scatterplot(x="windspeed",y = "cnt",data = data_vis,palette="Set3",ax=x[1][1])


# ## Outlier Analysis

# In[289]:


## Checking the presence of outlier in continous variables
sns.boxplot(data = data[['temp','atemp','windspeed','hum']])
fig  = plt.gcf()
fig.set_size_inches(8,8)


# In[290]:


## Removing outlier and checking correlation between target variable and independent continous variables

print(data.shape)
print(data['hum'].corr(data['cnt']))
print(data['windspeed'].corr(data['cnt']))

q75, q25 = np.percentile(data.loc[:,'hum'],[75,25])
iqr = q75 - q25

min = q25-(iqr*1.5)
max = q75+(iqr*1.5)

 
print(min)
print(max)

data = data.drop(data[data.loc[:,'hum']<min].index)
data = data.drop(data[data.loc[:,'hum']>max].index)


q75, q25 = np.percentile(data.loc[:,'windspeed'],[75,25])
iqr = q75 - q25

min = q25-(iqr*1.5)
max = q75+(iqr*1.5)

 
print(min)
print(max)

data = data.drop(data[data.loc[:,'windspeed']<min].index)
data = data.drop(data[data.loc[:,'windspeed']>max].index)


# ## Missing Value Analysis

# In[291]:


total = data.isnull().sum().sort_values(ascending=False)
percent = (data.isnull().sum()/data.isnull().count()).sort_values(ascending=False)
missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Percent'])
missing_data.head(30)

# There are no missing vlaues present after outlier analysis


# ## Feature Selection

# In[292]:


def Correlation(df):
    df_corr = df.loc[:,df.columns]
    corr = df_corr.corr()
    sns.set()
    plt.figure(figsize=(10,10))
    sns.heatmap(corr,annot=True,fmt=".3f",square=True,linewidths=0.5)

    
Correlation(data)


# In[293]:


## There is high correlation between temp and atemp variable
## there is very weak relation between holiday, weekday and working day variables
## So we will drop those variables
data_fs = data.drop(['atemp','holiday','weekday','workingday'],axis=1)
data_fs.head()


# In[294]:


# Splitting Dataset into train and test dataset

train,test = train_test_split(data_fs,test_size=0.2,random_state=121)


# ## Feature Scaling 

# In[295]:


## Data is normalized no need to do feature scaling
train.head()


# ## Error Metrics 

# In[296]:


## Defining Performance Metrics

def MAPE(y_true, y_pred):
    MAE = np.mean(np.abs((y_true - y_pred)))
    mape = np.mean(np.abs((y_true - y_pred) / y_true))
    print("MAE is:", MAE)
    print("MAPE is:", mape)
    return mape

def RMSE(y_true, y_pred):
    mse = np.mean((y_true - y_pred)**2)
    rmse = np.sqrt(mse)
    print("MSE: ",mse)
    print("RMSE: ",rmse)
    return rmse
    


# ## Linear Regression 

# In[297]:


LR_model = sm.OLS(train.iloc[:,7],train.iloc[:,0:6]).fit()

#Summary
print(LR_model.summary())

#Predict
LR_Model_predict = LR_model.predict(test.iloc[:,0:6])


# In[298]:


MAPE(test.iloc[:,7],LR_Model_predict)
RMSE(test.iloc[:,7],LR_Model_predict)


# In[299]:


result = pd.DataFrame({'Actual Value':test.iloc[:,7],'Linear Regression':LR_Model_predict})
result.head()


# ## Desicion Tree

# In[300]:


DT_model = DecisionTreeRegressor(random_state=100).fit(train.iloc[:,0:6],train.iloc[:,7])

#prediction
DT_model_predict = DT_model.predict(test.iloc[:,0:6],DT_model)


# In[ ]:





# In[301]:


MAPE(test.iloc[:,7],DT_model_predict)
RMSE(test.iloc[:,7],DT_model_predict)


# In[302]:


result['Desicion Tree'] = DT_model_predict
result.head()


# ## Random Forest

# In[303]:


RF_model = RandomForestRegressor(random_state=123)
np.random.seed(10)

arg_dict = {'max_depth':[2,4,6,8,10],
           'bootstrap':[True,False],
           'max_features':['auto','sqrt','log2',None],
           'n_estimators':[100,200,300,400,500]}

gs_randomForest = RandomizedSearchCV(RF_model,cv=10,param_distributions=arg_dict,
                                    n_iter=10)

gs_randomForest.fit(train.iloc[:,0:6],train.iloc[:,7])
print("Best Parameters using random Search",
     gs_randomForest.best_params_)


# In[304]:


RF_model.set_params(n_estimators = 500,
                   max_features='sqrt',
                   max_depth=8,
                   bootstrap=True)
RF_model.fit(train.iloc[:,0:6],train.iloc[:,7])
RF_model_predict = RF_model.predict(test.iloc[:,0:6])


# In[305]:


MAPE(test.iloc[:,7],RF_model_predict)
RMSE(test.iloc[:,7],RF_model_predict)


# In[306]:


result['Random Forest'] = RF_model_predict
result.head()


# From above models Random forest is performing well according to RMSE values

# In[307]:


#Saving the result of test data onto local machine
result.to_csv("Test_Result_python.csv",index=False)


# In[ ]:

File: Bike_Rental_Count.R
#Clean the environment

rm(list=ls())

##Set the working directory

setwd("D:/Amol_Data/Edwisor/Assignments/Project_2")
getwd()

##Load the required libraries

libraries = c("ggplot2","plyr","dplyr","rpart","gbm","DMwR","randomForest","usdm","corrgram","DataCombine")
lapply(X=libraries,require,character.only=TRUE)

#Load the dataset
data_day = read.csv(file="day.csv",header = T,sep=",",na.strings = c(" ","","NA"))


#######################Exploraty Data Analysis###############################################################

head(data_day)

dim(data_day)
str(data_day)

##Droping data which is not essential
#instant- index number
#dteday - all the required values are present
#casual,registered - cnt variable is  the sum of casual and registered variable

data_day <- subset(data_day,select = -c(instant,dteday,casual,registered))


data_day$season = as.factor(data_day$season)
data_day$yr =as.factor(data_day$yr)
data_day$mnth = as.factor(data_day$mnth)
data_day$weekday =as.factor(data_day$weekday)
data_day$workingday = as.factor(data_day$workingday)
data_day$holiday = as.factor(data_day$holiday)


###################################Univarent Analysis####################################

data_vis <- data_day

data_vis$season = factor( x = data_vis$season,level = c(1,2,3,4),labels = c("Spring","Summer","Fall","Winter"))
data_vis$yr = factor(x=data_vis$yr,levels = c(0,1),labels = c("2011","2012"))
data_vis$holiday = factor(x=data_vis$holiday,levels = c(0,1),labels = c("Working","Holiday"))
data_vis$weathersit =factor(x=data_vis$weathersit,levels = c(1,2,3,4),labels = c("Clear","Cloudy/Mist","Rain/Snow?Fog","Heavy Rain/Snow/Fog"))


bar1 = ggplot(data = data_vis,aes(x = season)) +geom_bar() + ggtitle("Season")
bar2 = ggplot(data = data_vis,aes(x=workingday)) + geom_bar() + ggtitle("working dayr")
bar3 = ggplot(data = data_vis,aes(x=holiday)) + geom_bar() + ggtitle("Holiday")
bar4 = ggplot(data = data_vis,aes(x=weathersit)) + geom_bar() + ggtitle("weather")

gridExtra::grid.arrange(bar1,bar2,bar3,bar4,ncol=2)

##################################### Bivarient Analysis########################################

sct1 = ggplot(data=data_vis,aes(x=temp,y=cnt)) + ggtitle("Temp") +geom_point() + xlab("Temperature") +ylab("Bike count")
sct2 = ggplot(data=data_vis,aes(x=atemp,y=cnt)) + ggtitle("aTemp") +geom_point() + xlab("Temperature") +ylab("Bike count")
sct3 = ggplot(data=data_vis,aes(x=hum,y=cnt)) + ggtitle("Humidity") +geom_point() + xlab("Humidity") +ylab("Bike count")
sct4 = ggplot(data=data_vis,aes(x=windspeed,y=cnt)) + ggtitle("Windspeed") +geom_point() + xlab("windspeed") +ylab("Bike count")
gridExtra::grid.arrange(sct1,sct2,sct3,sct4,ncol=2)


####################################BOx plot  for outlier analysis################################
cnames = colnames(data_day[,c("temp","atemp","hum","windspeed")])
cnames
 for( i in 1:length(cnames))
 {
   assign(paste0("gn",i),ggplot(aes_string(x=cnames[i],y='cnt'), data=data_day) +
            stat_boxplot(geom = "errorbar",width =0.5)+
            geom_boxplot(outlier.colour = "red",fill="grey",outlier.shape =18,
                         outlier.size=1,notch=FALSE)+
          theme(legend.position = "bottom")+
            labs(x=cnames[i])+
            ggtitle(paste("Box plot for",cnames[i])))
 }

gridExtra::grid.arrange(gn1,gn2,gn3,gn4,ncol=4)

cor(data_day$temp,data_day$cnt)
cor(data_day$hum,data_day$cnt)

##Remove outlier in windspeed

val = data_day$windspeed[data_day$windspeed %in% boxplot.stats(data_day$windspeed)$out]
val
data_day = data_day[which(!data_day$windspeed %in% val),]


val = data_day$hum[data_day$hum %in% boxplot.stats(data_day$hum)$out]
val
data_day = data_day[which(!data_day$hum %in% val),]


cnames = colnames(data_day[,c("hum","windspeed")])
cnames
for( i in 1:length(cnames))
{
  assign(paste0("gn",i),ggplot(aes_string(x=cnames[i],y='cnt'), data=data_day) +
           stat_boxplot(geom = "errorbar",width =0.5)+
           geom_boxplot(outlier.colour = "red",fill="grey",outlier.shape =18,
                        outlier.size=1,notch=FALSE)+
           theme(legend.position = "bottom")+
           labs(x=cnames[i])+
           ggtitle(paste("Box plot for",cnames[i])))
}

gridExtra::grid.arrange(gn1,gn2,ncol=2)

#####################################Missing Value Analysis#################################
missing_values = sapply(data_day,function(x){sum(is.na(x))})
print(missing_values)

## No missing Values

########################################### Feature scaling##################################

##In dataset numeric continous variables "temp","atemp","hum",and "windspeed" are in normalized form
## No need to feature scaling

########################################## Feature selection ################################

df_vif = data_day[,c("temp","atemp","hum","windspeed")]
vifcor(df_vif)


corrgram(data_day,order =F,upper.panel = panel.pie, text.panel = panel.txt,main="Correlation Plot")

names(data_day)
data_day <- subset(data_day,select = c("season","yr","mnth","weekday","temp","hum","windspeed","cnt"))

rmExcept(keepers = "data_day")


##Split data into train and test
set.seed(123)

train_index = sample(1:nrow(data_day),0.8*nrow(data_day))
train = data_day[train_index,]
test = data_day[-train_index,]



############################################   Linear Regression #############################

#Train
LR_model = lm(formula = cnt~.,data = train)

##summary

summary(LR_model)

LR_model_prediction = predict(LR_model,test[,-8])

df = data.frame("actual_value" =test[,8],"LR_model_predict"= LR_model_prediction)
head(df)


regr.eval(trues = test[,8],preds = LR_model_prediction,stats = c("mae","mse","rmse","mape"))

## mae          mse         rmse         mape 
## 6.422606e+02 7.366696e+05 8.582946e+02 2.171891e-01 


######################################Decision tree ################################

set.seed(12)
DT_model = rpart(cnt~.,data = train,method = "anova")

DT_model_predict = predict(DT_model,test[,-8])

df = data.frame(df,DT_model_predict)
head(df)
par(cex = 0.8)
plot(DT_model)
text(DT_model)


regr.eval(trues = test[,8],preds = DT_model_predict,stats = c("mae","mse","rmse","mape"))

## mae          mse         rmse         mape 
## 7.329756e+02 9.471663e+05 9.732247e+02 2.831340e-01 



##################################### Random Forest ##################################

RF_model = randomForest(cnt~.,data=train,ntree=500,nodesize=8,importance=TRUE)
RF_model_predict = predict(RF_model,test[,-8])
df = cbind(df,RF_model_predict)
head(df)


regr.eval(trues = test[,8],preds = RF_model_predict,stat = c("mae","mse","rmse","mape"))

## mae          mse         rmse         mape 
## 6.033788e+02 6.486310e+05 8.053763e+02 2.436907e-01 

############  Result   ##################
write.csv(df,"Test_result_R.csv",row.names = FALSE)
## From Above three model's RMSE metrics we can say that Random Forest works better than Decision Tree and Linear Regression